libpod/container_api.go

```diff
@@ -834,3 +840,181 @@ func (c *Container) Refresh(ctx context.Context) error {
 
 	return nil
 }
+
+// Checkpoint checkpoints a container
+func (c *Container) Checkpoint(keep bool) error {
+	if !c.batched {
+		c.lock.Lock()
+		defer c.lock.Unlock()
+
+		if err := c.syncContainer(); err != nil {
+			return err
+		}
+	}
+
+	if c.state.State != ContainerStateRunning {
+		return errors.Wrapf(ErrCtrStateInvalid, "%q is not running, cannot checkpoint", c.state.State)
+	}
+	if err := c.runtime.ociRuntime.checkpointContainer(c); err != nil {
+		return err
+	}
+
+	// Save network.status. This is needed to restore the container with
+	// the same IP. Currently limited to one IP address in a container
+	// with one interface.
+	formatJSON, err := json.MarshalIndent(c.state.NetworkStatus, "", "	")
+	if err != nil {
+		return err
+	}
+	if err := ioutil.WriteFile(filepath.Join(c.bundlePath(), "network.status"), formatJSON, 0644); err != nil {
+		return err
+	}
+
+	logrus.Debugf("Checkpointed container %s", c.ID())
+
+	c.state.State = ContainerStateStopped
+
+	// This needs to be deleted so that the restore works correctly
+	os.Remove(filepath.Join(c.bundlePath(), "attach"))
+	os.Remove(filepath.Join(c.bundlePath(), "ctl"))
+
+	// Cleanup Storage and Network
+	if err := c.cleanup(); err != nil {
+		return err
+	}
+
+	if keep {
+		return c.save()
+	}
+
+	// Remove log file
+	os.Remove(filepath.Join(c.bundlePath(), "dump.log"))
+	// Remove statistic file
+	os.Remove(filepath.Join(c.bundlePath(), "stats-dump"))
+
+	return c.save()
+}
+
+// Restore restores a container
+func (c *Container) Restore(keep bool) (err error) {
```

Comment:
Is there an easy way to check that a checkout exists at the start of the function, so we can error out if one doesn't? Should we track checkpoints in the database?

Reply:
I can test for the checkpoint directory and for one of the central checkpoint files. If both exist it should be a real checkpoint. Not sure if it should be tracked in the database.

Reply:
Testing for existence sounds sufficient. We might want to track them if we ever start adding more than one, but checking for existence is plenty good for now.